components/constellation/constellation.rs

```diff
@@ -2415,6 +2410,8 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
                 self.notify_history_changed(change.top_level_browsing_context_id);
             },
             Some(old_pipeline_id) => {
+                // https://html.spec.whatwg.org/multipage/#unload-a-document
+                self.unload_document(old_pipeline_id);
```

Comment:
@cbrewster Moving the unloading to `change_session_history` seems to work pretty well, with the exception of https://github.com/servo/servo/blob/d8dfa989704591cf6ec2417db6ee3e73530c71de/tests/wpt/web-platform-tests/html/browsers/browsing-the-web/unloading-documents/unload/003.html#L22 which now fails.
It seems that the tests expects the `unload` event to fire while the `iframe.contentDocument` would still be "003-1.html", while the session history is changed when it has already become"002-2.html".

Reply:
I think in this particular case we are unloading as part of [update the session history with the new page](https://html.spec.whatwg.org/multipage/browsing-the-web.html#update-the-session-history-with-the-new-page), and when this should occur is in fact quite vague per the spec, for example "After creating the Document object, but before any script execution, certainly before the parser stops, the user agent must update the session history with the new page." I do think we do it before the parser stops, since `change_session_history` is called in response to receiving a `ScriptMsg::ActivateDocument`, itself sent [here](https://github.com/servo/servo/blob/master/components/script/script_thread.rs#L2175) before the parser is even started. 

Perhaps the test makes too many assumptions about implementation details going beyond the spec?

Reply:
It is possible that the test is making too many assumptions. I'm okay with looking into this as a followup.